lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -1714,10 +1714,12 @@ bool SROAGlobalAndAllocas(HLModule &HLM, bool bHasDbgInfo) {
       sz0 = getNestedLevelInStruct(a0ty);
       sz1 = getNestedLevelInStruct(a1ty);
     }
-    // If sizes are equal, tiebreak with alphabetically lesser at higher priority
-    return sz0 < sz1 || (sz0 == sz1 && isa<GlobalVariable>(a0) &&
-                         isa<GlobalVariable>(a1) &&
-                         a0->getName() > a1->getName());
+    // If sizes are equal, and the new value is a GV,
+    // replace the existing node if it isn't GV or comes later alphabetically
+    // Thus, entries are sorted by size, global variableness, and then name
+    return sz0 < sz1 || (sz0 == sz1 && isa<GlobalVariable>(a1) &&
```

Comment:
This is pretty hacky. Can we file a bug tagged as `tech-debt` to revisit this and have a better way to maintain stable ordering? Respecting source ordering would be much better because it is at least intuitive.

Reply:
It's also pretty hard to read. Can we split up the logic or use `std::tie` to simplify? If I'm reading this correctly it should be equivalent to something like the following:
```c++
    return std::tie(sz0, isa<GlobalVariable>(a0), a0->getName()) <
           std::tie(sz1, isa<GlobalVariable>(a1), a1->getName());
```
That is, first compare on size, then sort non-GVs before GVs, then compare the names.

Reply:
Oh, we'd probably need a couple of variables to do this since `std::tie` needs lvalues, so more like this I guess, but you get the idea.
```c++
    bool A0IsGV = isa<GlobalVariable>(a0);
    bool A1IsGV = isa<GlobalVariable>(a1);
    StringRef A0Name = a0->getName();
    StringRef A1Name = a1->getName();
    return std::tie(sz0, A0IsGV, A0Name) < std::tie(sz1, A1IsGV, A1Name);
```

Reply:
Couldn't you use `std::make_tuple()` instead?